---
title: "MAS and control monocytes and lymphocytes, 10X study"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(dplyr)
library(ggplot2); theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill=NA, size=1),
          axis.text=element_text(colour="black"),
          axis.ticks=element_line(colour="black"),
          legend.key = element_blank(),
          strip.text.x = element_text(size = 14,margin = margin( b = 2, t = 2) ),
            strip.background = element_rect(fill="white", colour="black")))

library(ggthemes)
library(ggbeeswarm)
library(viridis)
library(stringr)
library(readxl)
library(kableExtra)
library(RColorBrewer)
library(plotly)
library(tidyr)
library(gtools)
library(data.table)
#library(edgeR)
library(ggrepel)
library(ComplexHeatmap)
library(egg) #For ggarrange
library(ggpubr) #Also for ggarrange
library(umap)
library(igraph)
library(forcats)
library(Seurat)
library(randomcoloR)
library(rcartocolor)  

opts_chunk$set(fig.width=6, fig.height=4.0, cache = TRUE, echo=FALSE, warning=FALSE, message=FALSE, cache.lazy = FALSE, results='hide')
opts_knit$set(root.dir = "/mnt/bioinformatics/workspace/hdeberg/P407_MASsJIA10X_CannyHamerman")

options(stringsAsFactors = FALSE)

options(future.globals.maxSize= 1591289600)


```

```{r set_up_directories, cache = TRUE}

baseDir <- "/mnt/bioinformatics/workspace/hdeberg/P407_MASsJIA10X_CannyHamerman"
dataDir <- file.path(baseDir, "data", "combined")
plotDir <- file.path(baseDir, "plots")
tableDir <- file.path(baseDir, "tables")

P407_1Samples <- c("1_mono_MAS_CIIT7181",
                   "2_mono_ctl",
                   "3_mono_MAS_CIIT7257",
                   "4_mono_ctl")

P407_1Files <- paste0("/mnt/bioinformatics/pipeline/Illumina/210901_VH00126_91_AAAH2HCHV/Project_P407-1Processed_bri_240202/",
                      P407_1Samples,
                      "/outs/per_sample_outs/",
                      P407_1Samples,
                      "/count/sample_filtered_feature_bc_matrix.h5")

P407_2Samples <- c("1_Monocyte",
                   "2_Monocyte",
                   "4_Monocyte",
                   "5_Monocyte",
                   "7_Monocyte",
                   "8_Monocyte",
                   "10_Monocyte",
                   "11_Monocyte",
                   "13_Monocyte")

P407_2LymphocyteSamples <- c("3_Lymphocyte",
                             "6_Lymphocyte",
                             "9_Lymphocyte",
                             "12_Lymphocyte")
  
P407_2Files <- paste0("/mnt/bioinformatics/pipeline/Illumina/230515_VH00126_276_AACH75YM5/Project_P407-2Processed_bri_240202/",
                      P407_2Samples,
                      "/outs/per_sample_outs/",
                      P407_2Samples,
                      "/count/sample_filtered_feature_bc_matrix.h5")


P407_2LymphocyteFiles <- paste0("/mnt/bioinformatics/pipeline/Illumina/230515_VH00126_276_AACH75YM5/Project_P407-2Processed_bri_240202/",
                      P407_2LymphocyteSamples,
                      "/outs/per_sample_outs/",
                      P407_2LymphocyteSamples,
                      "/count/sample_filtered_feature_bc_matrix.h5")




```

```{r read_P407_1}

processP407_1 <- function(sample){
  print(sample) #debugging 

  gexAb <- Read10X_h5(sample)
  
  seuratSample <- CreateSeuratObject(counts = gexAb,
                                     min.features = 100,
                                     min.cells = 3)
  
  
  #Clean memory
  remove(gexAb)
  
  seuratSample[["sample"]] <- str_extract(sample, "[0-9]_mono")
  seuratSample[["percentMT"]] <- PercentageFeatureSet(seuratSample, pattern = "^MT-")
  
  return(seuratSample)

}

gexMonoP407_1 <- lapply(P407_1Files, processP407_1)

```

```{r read_p407_2}

processP407_2 <- function(sample){
  print(sample) #debugging 

  gexAb <- Read10X_h5(sample)
  
  seuratSample <- CreateSeuratObject(counts = gexAb$`Gene Expression`,
                                                        min.features = 100,
                                                        min.cells = 3)
  
  abData <- gexAb$`Antibody Capture`
  
  keepCells <- colnames(seuratSample)
  abData <- abData[,keepCells]
  
  #Clean memory
  remove(gexAb)
  
  seuratSample[["sample"]] <- str_extract(sample, "[0-9]+_Mono|[0-9]+_Lymph")
  seuratSample[["percentMT"]] <- PercentageFeatureSet(seuratSample, pattern = "^MT-")
  
  return(seuratSample)

}

processP407_2Lymph <- function(sample){
  print(sample) #debugging 

  gexAb <- Read10X_h5(sample)
  
  seuratSample <- CreateSeuratObject(counts = gexAb$`Gene Expression`,
                                                        min.features = 100,
                                                        min.cells = 3)
  
  abData <- gexAb$`Antibody Capture`
  
  keepCells <- colnames(seuratSample)
  abData <- abData[,keepCells]
  
  #Clean memory
  remove(gexAb)
  
  seuratSample[["sample"]] <- str_extract(sample, "[0-9]+_Lymph")
  seuratSample[["percentMT"]] <- PercentageFeatureSet(seuratSample, pattern = "^MT-")
  
  seuratSample[["HT"]] <- CreateAssayObject(abData)
  
  return(seuratSample)

}

gexMonoP407_2 <- lapply(P407_2Files, processP407_2)
gexLymphP407_2 <- lapply(P407_2LymphocyteFiles, processP407_2Lymph)

```

```{r merge}


mergedMonoP407_1 <- merge(gexMonoP407_1[[1]], 
                       y = c(gexMonoP407_1[2:length(gexMonoP407_1)]),
                       merge.data = TRUE)

mergedMonoP407_2 <- merge(gexMonoP407_2[[1]], 
                       y = c(gexMonoP407_2[2:length(gexMonoP407_2)]),
                       merge.data = TRUE)

mergedLymph <- merge(gexLymphP407_2[[1]], 
                       y = c(gexLymphP407_2[2:length(gexLymphP407_2)]),
                       merge.data = TRUE)

#clean
remove(gexMonoP407_1)
remove(gexMonoP407_2)
remove(gexLymphP407_2)

mergedMono <- merge(mergedMonoP407_1,
                    mergedMonoP407_2)

remove(mergedMonoP407_1)
remove(mergedMonoP407_2)
```

```{r add_anno}

aggCSV <- read.csv(file.path("/mnt/bioinformatics/pipeline/tenxSummaries/230720-P407-1and2", 
                             "aggregated",
                             "P407-1and2_aggr_aggregation.csv"))
aggCSV$suffix <- rownames(aggCSV)

dataDir <- file.path(baseDir, "data", "combined")

load(file.path(dataDir,"P407_2Anno.Rdata"))

anno407_1 <- read_excel(file.path(dataDir, "P407-1_Final Annotation.xlsx"),
                        "Sheet1")

anno <- bind_rows(anno407_1,
                  anno407_2)

anno$sampleName <- coalesce(anno$sample_name,
                            anno$`Sample Name`)

anno$cellType <- coalesce(anno$cellType,
                          anno$`Content Type`)

anno$cellNumber <- coalesce(anno$cellNumber,
                          anno$`Cell Number`)

anno$project <- coalesce(anno$project,
                         anno$`Proj-Sub...1`)

anno$donorID <- coalesce(anno$donorID,
                         anno$`Donor ID`)

anno$sex <- coalesce(anno$sex,
                     anno$Sex)

anno$studyGroup <- coalesce(anno$studyGroup,
                            anno$`Study Group`)

#Make sample names match in anno and seuratObj 13 is a redo of 8
anno$sampleName[anno$sampleName == "13_Monocyte_8redo"] <- "13_Monocyte"

mergedMono$sampleName <- case_when(mergedMono$sample == "1_mono" ~ "1_mono_MAS_CIIT7181",
                                   mergedMono$sample == "2_mono" ~ "2_mono_ctl",
                                   mergedMono$sample == "3_mono" ~ "3_mono_MAS_CIIT7257",
                                   mergedMono$sample == "4_mono" ~ "4_mono_ctl",
                                   mergedMono$sample == "1_Mono" ~ "1_Monocyte",
                                   mergedMono$sample == "2_Mono" ~ "2_Monocyte",
                                   mergedMono$sample == "4_Mono" ~ "4_Monocyte",
                                   mergedMono$sample == "5_Mono" ~ "5_Monocyte",
                                   mergedMono$sample == "7_Mono" ~ "7_Monocyte",
                                   mergedMono$sample == "8_Mono" ~ "8_Monocyte",
                                   mergedMono$sample == "10_Mono" ~ "10_Monocyte",
                                   mergedMono$sample == "11_Mono" ~ "11_Monocyte",
                                   mergedMono$sample == "13_Mono" ~ "13_Monocyte")

mergedLymph$sampleName <- case_when(mergedLymph$sample == "3_Lymph" ~ "3_Lymphocyte",
                                    mergedLymph$sample == "6_Lymph" ~ "6_Lymphocyte",
                                    mergedLymph$sample == "9_Lymph" ~ "9_Lymphocyte",
                                    mergedLymph$sample == "12_Lymph" ~ "12_Lymphocyte")

mergedMono$project <- anno$project[match(mergedMono$sampleName, anno$sampleName)]
mergedLymph$project <- anno$project[match(mergedLymph$sampleName, anno$sampleName)]

```

```{r qc_nfeature}

#QC cutoffs should be applied differently in P407-1 and P407-2

#Select QC cutoffs separately for -1 and -2

nFeatureLow1 <-500 
nFeatureHigh1 <- 4000
pctMtCutoff1 <- 15

nFeatureLow2 <-500 
nFeatureHigh2 <- 2500
pctMtCutoff2 <- 5

mergedMono$qcPass <- case_when(mergedMono$project == "P407-1" ~ (mergedMono$nFeature_RNA > nFeatureLow1 & mergedMono$nFeature_RNA < nFeatureHigh1 & mergedMono$percentMT < pctMtCutoff1), 
                               mergedMono$project == "P407-2" ~ (mergedMono$nFeature_RNA > nFeatureLow2 & mergedMono$nFeature_RNA < nFeatureHigh2 & mergedMono$percentMT < pctMtCutoff2))

mergedLymph$qcPass <- mergedLymph$nFeature_RNA > nFeatureLow2 & mergedLymph$nFeature_RNA < nFeatureHigh2 & mergedLymph$percentMT < pctMtCutoff2

seuratMonoQC <- subset(mergedMono, 
                   subset = (qcPass == TRUE))

seuratLymphQC <- subset(mergedLymph, 
                   subset = (qcPass == TRUE))

remove(mergedMono)
remove(mergedLymph)

```

# Monocyte processing

```{r annotate_mono}

seuratMonoQC$donorID <- anno$donorID[match(seuratMonoQC$sampleName, anno$sampleName)]
seuratMonoQC$sex <- anno$sex[match(seuratMonoQC$sampleName, anno$sampleName)]

seuratMonoQC$sex <- case_when(seuratMonoQC$donorID %in% c("CIIT7174",
                                                          "CIIT7193",
                                                          "sJIA001",
                                                          "sJIA007",
                                                          "xbp657374814") ~ "female",
                              seuratMonoQC$donorID %in% c("CIIT7344",
                                                          "sJIA002",
                                                          "xbp425617024") ~ "male",
                              TRUE ~ anno$sex[match(seuratMonoQC$sampleName, anno$sampleName)])

seuratMonoQC$sex[seuratMonoQC$sex == "M"] <- "male"
seuratMonoQC$sex[seuratMonoQC$sex == "F"] <- "female"

seuratMonoQC$age <- case_when(seuratMonoQC$donorID == "CIIT7174" ~ 16,
                              seuratMonoQC$donorID == "CIIT7193" ~ 5,
                                seuratMonoQC$donorID == "sJIA001" ~ 12,
                                seuratMonoQC$donorID == "sJIA007" ~ 2,
                                seuratMonoQC$donorID == "xbp657374814" ~ 13,
                                seuratMonoQC$donorID == "CIIT7344" ~ 0.9,
                                seuratMonoQC$donorID == "sJIA002" ~ 7,
                                seuratMonoQC$donorID == "xbp425617024" ~ 6,
                              TRUE ~ anno$age[match(seuratMonoQC$sampleName, anno$sampleName)])

seuratMonoQC$studyGroup <- case_when(seuratMonoQC$donorID %in% c("CIIT7174",
                                                          "CIIT7193",
                                                          "sJIA001",
                                                          "sJIA007",
                                                          "CIIT7344",
                                                          "sJIA002") ~ "MAS",
                              seuratMonoQC$donorID %in% c("xbp657374814",
                                                          "xbp425617024") ~ "HC",
                              TRUE ~ anno$studyGroup[match(seuratMonoQC$sampleName, anno$sampleName)])

seuratMonoQC$studyGroup[seuratMonoQC$studyGroup == "control"] <- "HC"


```

```{r split_process_mono}

seuratMonoQC <- JoinLayers(seuratMonoQC)

seuratMonoQC[["RNA"]] <- split(seuratMonoQC[["RNA"]], f = seuratMonoQC$donorID)

#This will normalize and find variable features for each layer separately. A consensus set of variable features is automatically identified by Seurat
seuratMonoQC <- NormalizeData(seuratMonoQC)
seuratMonoQC <- FindVariableFeatures(seuratMonoQC)
seuratMonoQC <- ScaleData(seuratMonoQC)
seuratMonoQC <- RunPCA(seuratMonoQC)

```

```{r integrate_mono}

seuratMonoQC <- IntegrateLayers(
  object = seuratMonoQC, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.rpca",
  verbose = FALSE
)

```

```{r process_integration}

# re-join layers after integration
seuratMonoQC[["RNA"]] <- JoinLayers(seuratMonoQC[["RNA"]])

seuratMonoQC <- FindNeighbors(seuratMonoQC, reduction = "integrated.rpca", dims = 1:30)
seuratMonoQC <- FindClusters(seuratMonoQC)

seuratMonoQC <- RunUMAP(seuratMonoQC, dims = 1:30, reduction = "integrated.rpca")

saveRDS(seuratMonoQC, file.path(dataDir, "SeuratV5MonoQC.RDS"))

seuratMonoQC <- readRDS(file.path(dataDir, "SeuratV5MonoQC.RDS"))

```

# Lymphocyte processing

```{r process_lymph}

seuratLymphQC <- NormalizeData(object = seuratLymphQC, 
                                    verbose = FALSE,
                                    normalization.method = "CLR", 
                                    margin = 2, 
                                    assay = "HT")

DefaultAssay(seuratLymphQC) <- "RNA"

seuratLymphQC <- NormalizeData(object = seuratLymphQC, verbose = FALSE)
seuratLymphQC <- FindVariableFeatures(object = seuratLymphQC, 
        selection.method = "vst", nfeatures = 2000, verbose = FALSE)

seuratLymphQC <- ScaleData(object = seuratLymphQC, verbose = FALSE)
seuratLymphQC <- RunPCA(object = seuratLymphQC, npcs = 30, verbose = FALSE)

```

```{r hashtag_demultiplexing}

#Check that cutoffs previously picked are still okay on merged object

htVariables <- c("C0259", "C0263")
DefaultAssay(seuratLymphQC) <- "HT"
htDF <- FetchData(seuratLymphQC,
                  vars = c(htVariables, "sample"))

C0259Histogram <- htDF %>%
  ggplot(aes(x = `C0259`))+
  geom_histogram()+
  labs(x = "C0259 expression",
       y = "Number of cells")

C0263Histogram <- htDF %>%
  ggplot(aes(x = `C0263`))+
  geom_histogram()+
  labs(x = "C0263 expression",
       y = "Number of cells")


#Set hashtag cutoffs and use these to remove doublets
C0259Cutoff <- 1
C0263Cutoff <- 1

seuratLymphQC$C0259 <-  htDF$C0259
seuratLymphQC$C0263 <-  htDF$C0263

seuratLymphQC$C0259Level <- case_when(seuratLymphQC$C0259 > C0259Cutoff ~ "C0259pos",
                               seuratLymphQC$C0259 <= C0259Cutoff ~ "C0259neg")

seuratLymphQC$C0263Level <- case_when(seuratLymphQC$C0263 > C0263Cutoff ~ "C0263pos",
                               seuratLymphQC$C0263 <= C0263Cutoff ~ "C0263neg")

#Check overlaps of high calls
table(seuratLymphQC$C0259Level, seuratLymphQC$C0263Level, useNA = "ifany")

seuratLymphQC@meta.data <- seuratLymphQC@meta.data %>%
  dplyr::mutate(manualDemultiplexing = case_when(C0259Level == "C0259pos" & C0263Level == "C0263pos" ~ "Doublet",
                                                 C0259Level == "C0259pos" & C0263Level == "C0263neg" ~ "C0259",
                                                 C0259Level == "C0259neg" & C0263Level == "C0263pos" ~ "C0263",
                                                 C0259Level == "C0259neg" & C0263Level == "C0263neg" ~ "Negative"))


#Use pos/neg to define lymphocute donorIDs

seuratLymphQC@meta.data <- seuratLymphQC@meta.data %>%
  dplyr::mutate(donorID = case_when(sampleName == "3_Lymphocyte" & manualDemultiplexing == "C0259"  ~ "CIIT7174",
                                    sampleName == "3_Lymphocyte" & manualDemultiplexing == "C0263"  ~ "CIIT7344",
                                    sampleName == "6_Lymphocyte" & manualDemultiplexing == "C0259"  ~ "xbp425617024",
                                    sampleName == "6_Lymphocyte" & manualDemultiplexing == "C0263"  ~ "sJIA002",
                                    sampleName == "9_Lymphocyte" & manualDemultiplexing == "C0259"  ~ "CIIT7193",
                                    sampleName == "9_Lymphocyte" & manualDemultiplexing == "C0263"  ~ "sJIA007",
                                    sampleName == "12_Lymphocyte" & manualDemultiplexing == "C0259"  ~ "xbp657374814",
                                    sampleName == "12_Lymphocyte" & manualDemultiplexing == "C0263"  ~ "sJIA001",
                                    manualDemultiplexing == "Negative" ~ "NegativeLymphocytes"
  ))

#Check 
#table(seuratLymphQC$donorID,useNA = "ifany")
seuratLymphQC <- subset(seuratLymphQC,
                       subset = (donorID != "NegativeLymphocytes"))


```

```{r add_pt_level_anno_lymph}

seuratLymphQC$sex <- case_when(seuratLymphQC$donorID %in% c("CIIT7174",
                                                          "CIIT7193",
                                                          "sJIA001",
                                                          "sJIA007",
                                                          "xbp657374814") ~ "female",
                              seuratLymphQC$donorID %in% c("CIIT7344",
                                                          "sJIA002",
                                                          "xbp425617024") ~ "male",
                              TRUE ~ anno$sex[match(seuratLymphQC$sampleName, anno$sampleName)])

seuratLymphQC$sex[seuratLymphQC$sex == "M"] <- "male"
seuratLymphQC$sex[seuratLymphQC$sex == "F"] <- "female"

seuratLymphQC$age <- case_when(seuratLymphQC$donorID == "CIIT7174" ~ 16,
                              seuratLymphQC$donorID == "CIIT7193" ~ 5,
                                seuratLymphQC$donorID == "sJIA001" ~ 12,
                                seuratLymphQC$donorID == "sJIA007" ~ 2,
                                seuratLymphQC$donorID == "xbp657374814" ~ 13,
                                seuratLymphQC$donorID == "CIIT7344" ~ 0.9,
                                seuratLymphQC$donorID == "sJIA002" ~ 7,
                                seuratLymphQC$donorID == "xbp425617024" ~ 6,
                              TRUE ~ anno$age[match(seuratLymphQC$sampleName, anno$sampleName)])

seuratLymphQC$studyGroup <- case_when(seuratLymphQC$donorID %in% c("CIIT7174",
                                                          "CIIT7193",
                                                          "sJIA001",
                                                          "sJIA007",
                                                          "CIIT7344",
                                                          "sJIA002") ~ "MAS",
                              seuratLymphQC$donorID %in% c("xbp657374814",
                                                          "xbp425617024") ~ "HC",
                              TRUE ~ anno$studyGroup[match(seuratLymphQC$sampleName, anno$sampleName)])

seuratLymphQC$studyGroup[seuratLymphQC$studyGroup == "control"] <- "HC"

```

```{r split_process_lymph}

DefaultAssay(seuratLymphQC) <-"RNA"

seuratLymphQC <- JoinLayers(seuratLymphQC)

seuratLymphQC[["RNA"]] <- split(seuratLymphQC[["RNA"]], f = seuratLymphQC$donorID)

#This will normalize and find variable features for each layer separately. A consensus set of variable features is automatically identified by Seurat
seuratLymphQC <- NormalizeData(seuratLymphQC)
seuratLymphQC <- FindVariableFeatures(seuratLymphQC)
seuratLymphQC <- ScaleData(seuratLymphQC)
seuratLymphQC <- RunPCA(seuratLymphQC)

```

```{r integrate_mono}

seuratLymphQC <- IntegrateLayers(
  object = seuratLymphQC, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.rpca",
  verbose = FALSE
)

```

```{r process_integration}

# re-join layers after integration
seuratLymphQC[["RNA"]] <- JoinLayers(seuratLymphQC[["RNA"]])

seuratLymphQC <- FindNeighbors(seuratLymphQC, reduction = "integrated.rpca", dims = 1:30)
seuratLymphQC <- FindClusters(seuratLymphQC)

seuratLymphQC <- RunUMAP(seuratLymphQC, dims = 1:30, reduction = "integrated.rpca")

saveRDS(seuratLymphQC, file.path(dataDir, "SeuratV5LymphQC.RDS"))

```

```{r umap_plots_lymph_qc}

DimPlot(seuratLymphQC, reduction = "umap", group.by = c("seurat_clusters")) 

```

```{r ref_mapping}

#Load reference data. Takes a few minutes
data10xMultimodalReference <-
  SeuratDisk::LoadH5Seurat(
    file.path(baseDir,
              "multimodalRefData",
              "pbmc_multimodal.h5seurat"))

#Takes 2-3 min

transferAnchorsData10xMultimodalReferenceLymph <- FindTransferAnchors(
  reference = data10xMultimodalReference,
  query = seuratLymphQC,
  normalization.method = "SCT",
  reference.reduction = "spca",
  dims = 1:50
)

seuratLymphQC <- MapQuery(
  anchorset = transferAnchorsData10xMultimodalReferenceLymph,
  query = seuratLymphQC,
  reference = data10xMultimodalReference,
  refdata = list(
    celltype.l1 = "celltype.l1",
    celltype.l2 = "celltype.l2",
    celltype.l3 = "celltype.l3",
    predicted_ADT = "ADT"
  ),
  reference.reduction = "spca",
  reduction.model = "wnn.umap"
)

transferAnchorsData10xMultimodalReferenceMono <- FindTransferAnchors(
  reference = data10xMultimodalReference,
  query = seuratMonoQC,
  normalization.method = "SCT",
  reference.reduction = "spca",
  dims = 1:50
)

seuratMonoQC <- MapQuery(
  anchorset = transferAnchorsData10xMultimodalReferenceMono,
  query = seuratMonoQC,
  reference = data10xMultimodalReference,
  refdata = list(
    celltype.l1 = "celltype.l1",
    celltype.l2 = "celltype.l2",
    celltype.l3 = "celltype.l3",
    predicted_ADT = "ADT"
  ),
  reference.reduction = "spca",
  reduction.model = "wnn.umap"
)

#saveRDS(seuratLymphQC, file.path(dataDir, "SeuratV5LymphQC.RDS"))
#saveRDS(seuratMonoQC, file.path(dataDir, "SeuratV5MonoQC.RDS"))

#Clean up

remove(data10xMultimodalReference)
remove(transferAnchorsData10xMultimodalReference)

#seuratLymphQC <- readRDS(file.path(dataDir, "SeuratV5LymphQC.RDS"))
#seuratMonoQC <- readRDS(file.path(dataDir, "SeuratV5MonoQC.RDS"))

```

# Monocyte analysis

```{r umap_plots_mono_qc}

DimPlot(seuratMonoQC, reduction = "umap", group.by = c("seurat_clusters"))

DimPlot(seuratMonoQC, reduction = "umap", group.by = c("predicted.celltype.l2"))

```

```{r subset_mono}

seuratMonoQC$UMAP1 <- seuratMonoQC[["umap"]]@cell.embeddings[,1]
seuratMonoQC$UMAP2 <- seuratMonoQC[["umap"]]@cell.embeddings[,2]

seuratMonoQC <- subset(seuratMonoQC,
                          subset = (UMAP1 > -10) & (UMAP2 < 6))

seuratMonoQC <- FindNeighbors(seuratMonoQC, reduction = "integrated.rpca", dims = 1:30)
seuratMonoQC <- FindClusters(seuratMonoQC, resolution = 0.3)

seuratMonoQC <- RunUMAP(seuratMonoQC, dims = 1:30, reduction = "integrated.rpca")

saveRDS(seuratMonoQC, file.path(dataDir, "SeuratV5MonoQC.RDS"))
seuratMonoQC <- readRDS(file.path(dataDir, "SeuratV5MonoQC.RDS"))

```

```{r umap_plots_mono_qc_again}
#Cluster colors

clusterColors <- rcartocolor::carto_pal(length(unique(seuratMonoQC$seurat_clusters)), "Vivid")
names(clusterColors) <- gtools::mixedsort(unique(seuratMonoQC$seurat_clusters))

DimPlot(seuratMonoQC, reduction = "umap", group.by = c("seurat_clusters"))

DimPlot(seuratMonoQC, reduction = "umap", group.by = c("seurat_clusters")) +
  facet_wrap(~seurat_clusters)

png(file.path(plotDir, "UMAPClustersRes0_3.png"),
    height = 400,
    width = 600)

DimPlot(seuratMonoQC, reduction = "umap", 
        group.by = c("seurat_clusters"),
        label = TRUE)+
  labs(x = "UMAP 1",
       y = "UMAP 2")+
  scale_color_manual(values = clusterColors) +
  theme(aspect.ratio = 1)

invisible(dev.off())

png(file.path(plotDir, "UMAP_MarkerGenesRes0_3.png"),
    height = 400,
    width = 600)

FeaturePlot(seuratMonoQC, reduction = "umap", 
        features = c("CD14", "FCGR3A", "CD1C", "PF4"))+
  labs(x = "UMAP 1",
       y = "UMAP 2")&
  theme(aspect.ratio = 1)

invisible(dev.off())


```


```{r cluster_de}

Idents(seuratMonoQC) <- "seurat_clusters"

DefaultAssay(seuratMonoQC) <- "RNA"

topMarkersSeuratMonoQC <- FindAllMarkers(seuratMonoQC)

write.csv(topMarkersSeuratMonoQC, file.path(tableDir, "SeuratV5LongDEGenesByClusterRNA_Res0_3.csv"),
          row.names = F,
          quote= F)

#Write to file
topMarkersList <- NULL
for(selCluster in unique(topMarkersSeuratMonoQC$cluster)){
  topMarkersList[[paste0("Cluster", selCluster)]] <- topMarkersSeuratMonoQC %>%
    dplyr::filter(cluster == selCluster) %>%
    dplyr::arrange(p_val_adj,
                   -1*avg_log2FC)
  
}

openxlsx::write.xlsx(topMarkersList, file = file.path(tableDir, "SeuratV5ClusterGenesRes0_3.xlsx"))

```

```{r top_cluster_genes}

topMarkers <- topMarkersSeuratMonoQC %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=10) %>%
  dplyr::slice_max(avg_log2FC, n = 10) %>%
  dplyr::mutate(fc_order = order(avg_log2FC)) 

topMarkers <- topMarkers %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")


topMarkers2 <- topMarkersSeuratMonoQC %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0.5,
                p_val_adj < 0.05,
                pct.1 > 0.2) %>%
  dplyr::filter(avg_log2FC > 0) %>%
  dplyr::slice_min(p_val_adj, n=20) %>%
  dplyr::slice_max(avg_log2FC, n = 20) %>%
  dplyr::mutate(fc_order = order(avg_log2FC,decreasing = T)) 

topMarkers2 <- topMarkers2 %>%
  pivot_wider(id_cols = "fc_order",
              names_from = "cluster",
              values_from = "gene")

write.csv(topMarkers2,
          file.path(tableDir, "TopMarkersRes0_3.csv"),
          row.names = F,
          quote = F)

nGenes <- 6

hmapGenes <- topMarkersSeuratMonoQC %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0.5,
                p_val_adj < 0.05,
                pct.1 > 0.2) %>%
  dplyr::slice_min(p_val_adj, n=nGenes) %>%
  dplyr::slice_max(avg_log2FC, n = nGenes) #%>%
  #dplyr::arrange(p_val_adj) %>%
  #dplyr::mutate(fc_order = order(p_val_adj))

selGenes <- unique(hmapGenes$gene)


hmap <- DoHeatmap(subset(seuratMonoQC, downsample = 100), 
                  features = selGenes, 
                  group.colors = clusterColors,
                  size = 3) +
  scale_fill_gradientn(colors = rcartocolor::carto_pal(5, "Geyser"))

png(file.path(plotDir, "HeatmapTop5GenesRes0_3.png"),
    height = 600,
    width = 600)

print(hmap)

invisible(dev.off())


```

```{r metacluster}

clusterTree <- BuildClusterTree(object = seuratMonoQC,
                                features = selGenes)
PlotClusterTree(clusterTree)

pdf(file.path(plotDir, "SeuratV5_Dendrogram_MonocytesInt_ClusterTree.pdf"),
    height = 4, 
    width = 8,
    onefile = F)

PlotClusterTree(clusterTree)

invisible(dev.off())


```

```{r barplots_cluster_proportions}

#For each donor plot the proportions of monocytes that fall into each subpopulation

totalCellCounts <- seuratMonoQC@meta.data %>%
  dplyr::group_by(donorID) %>%
  dplyr::summarise(nCellsTot = n(),
                   studyGroup = unique(studyGroup)) 

donorMonoClusterCounts <- seuratMonoQC@meta.data %>%
  dplyr::group_by(donorID, seurat_clusters) %>%
  dplyr::summarise(nCells= n()) %>%
  dplyr::ungroup() %>%
  complete(donorID, seurat_clusters, fill = list(nCells = 0))

donorMonoClusterCounts <- left_join(donorMonoClusterCounts, totalCellCounts)

donorMonoClusterCounts <- donorMonoClusterCounts%>%
  dplyr::mutate(propCells = nCells/nCellsTot)

gMonoPopProportions <- donorMonoClusterCounts %>%
  ggplot(aes(x=donorID, 
             y=propCells, 
             fill=seurat_clusters)) + 
  geom_col() + 
  labs(x = "Donor",
       y = "Fraction of cells in donor",
       fill = "")+ 
  scale_fill_manual(values = clusterColors) +
  scale_y_continuous(labels = scales::percent) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

pdf(file.path(plotDir, "BarPlots_MonocytesInt_seurat_clusterss.pdf"),
    height = 6, 
    width = 8)

print(gMonoPopProportions)

invisible(dev.off())

#Boxplots of proportions

gMonoPopProportionsBox <- donorMonoClusterCounts %>%
  ggplot(aes(x=studyGroup, 
             y=propCells, 
             color=seurat_clusters)) + 
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom() +
  labs(x = "",
       y = "Fraction of cells in donor",
       fill = "")+ 
  scale_color_manual(values = clusterColors) +
  scale_y_continuous(labels = scales::percent) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~seurat_clusters,
             scales = "free_y")

pdf(file.path(plotDir, "BoxPlots_MonocytesIntIntegrated_seurat_clusterss.pdf"),
    height = 6, 
    width = 8)

print(gMonoPopProportionsBox)

invisible(dev.off())

#Heatmap of proportions

hmapProp <- donorMonoClusterCounts %>%
  pivot_wider(id_cols = "donorID",
              names_from = "seurat_clusters",
              values_from = "propCells") %>%
  as.data.frame()

rownames(hmapProp) <- hmapProp$donorID
hmapProp$donorID <- NULL

h <- Heatmap(hmapProp,
              name = "fraction cells in donor")

draw(h)

pdf(file.path(plotDir, "Heatmap_MonocytesIntIntegrated_seurat_clustersProportions.pdf"),
    height = 4, 
    width = 8)

draw(h)

invisible(dev.off())

#Write out proportions

write.csv(donorMonoClusterCounts,
          file = file.path(tableDir, "SeuratV5ClusterProporitonsLongFormat.csv"),
          row.names = F,
          quote = F)

write.csv(hmapProp,
          file = file.path(tableDir, "SeuratV5ClusterProporitonsWideFormat.csv"),
          row.names = T,
          quote = F)

```

```{r cluster_stats}

clusterStats <- donorMonoClusterCounts %>%
  dplyr::group_by(seurat_clusters) %>%
  rstatix::t_test(propCells ~ studyGroup)

```

```{r plot_key_cluster_genes}

png(file.path(plotDir, "UMAP_AlarminGenes_Monocytes.png"),
    height = 400,
    width = 600)

FeaturePlot(seuratMonoQC, reduction = "umap", 
        features = c("S100A8", "S100A9", "S100A12"))+
  labs(x = "UMAP 1",
       y = "UMAP 2")&
  scale_color_viridis()&
  theme(aspect.ratio = 1)

invisible(dev.off())

```

```{r compare_cd14_mono_clusters}

Idents(seuratMonoQC) <- "seurat_clusters"

DefaultAssay(seuratMonoQC) <- "RNA"

topMarkersClust0 <- FindMarkers(seuratMonoQC,
                                  ident.1 = 0,
                                  ident.2 = c(1,3))

topMarkersClust1 <- FindMarkers(seuratMonoQC,
                                  ident.1 = 1,
                                  ident.2 = c(0,3))

topMarkersClust3 <- FindMarkers(seuratMonoQC,
                                  ident.1 = 3,
                                  ident.2 = c(0,1))

topGenesClust0 <- topMarkersClust0 %>%
  dplyr::filter(pct.1 > 0.5)

topGenesClust1 <- topMarkersClust1 %>%
  dplyr::filter(pct.1 > 0.5)

topGenesClust3 <- topMarkersClust3 %>%
  dplyr::filter(pct.1 > 0.5)

png(file.path(plotDir, "UMAP_MarkerGenesCD14Clustsers_0_1_3.png"),
    height = 900,
    width = 800)

FeaturePlot(seuratMonoQC, reduction = "umap", 
        features = c("NFKB1","PID1", "SMAD3",
                     "S100A12", "S100A8", "GAPDH",
                     "ACSL1", "MALAT1", "NAMPT"),
        ncol = 3)+
  labs(x = "UMAP 1",
       y = "UMAP 2")&
  scale_color_viridis() &
  theme(aspect.ratio = 1)

invisible(dev.off())

```

```{r bulk_rnaseq_signatures}

bulkMASvsHC <- read.csv(file.path(baseDir,
                                  "tables",
                                  "bulk",
                                  "MASvsHCGeneStatistics.csv"))

bulkSigGenes <- bulkMASvsHC$Gene[(abs(bulkMASvsHC$log2FoldChange) > 1) &
                                   (bulkMASvsHC$MultipleTestingCorrectedFDR < 0.1)]

masUp <- bulkSigGenes[bulkSigGenes %in% bulkMASvsHC$Gene[bulkMASvsHC$log2FoldChange>0]]
masDn <- bulkSigGenes[bulkSigGenes %in% bulkMASvsHC$Gene[bulkMASvsHC$log2FoldChange<0]]

bulkSigGenes <- bulkSigGenes[bulkSigGenes %in% rownames(seuratMonoQC)]

Idents(seuratMonoQC) <- "seurat_clusters"

hmap <- DoHeatmap(subset(seuratMonoQC, downsample = 100), 
                  features = bulkSigGenes, 
                  size = 3,
                  group.colors = clusterColors) &
  scale_fill_gradientn(colors = rcartocolor::carto_pal(5, "Geyser"))
png(file.path(plotDir,
              "Heatmap_Monocytes_BulkDEGenes.png"),
    height = 1200,
    width = 800)

print(hmap)

invisible(dev.off())


seuratMonoQC$geneSetScore <- (rowSums(FetchData(seuratMonoQC,masUp)) - rowSums(FetchData(seuratMonoQC,masDn)))/length(bulkSigGenes)

gBulkSig <- FeaturePlot(seuratMonoQC,
            features = "geneSetScore",
            reduction = "umap", 
            split.by = "studyGroup",
            ncol = 2,
            min.cutoff = -0.1,
            max.cutoff = 0.2,
        raster = FALSE)&
  labs(x = "UMAP 1", y ="UMAP 2", color = "") &
  scale_color_viridis() &
  theme(aspect.ratio = 1) & theme(legend.position = "right")

png(file.path(plotDir,
              "UMAP_Monocytes_BulkSig.png"),
    height = 400,
    width = 1200)

print(gBulkSig)

invisible(dev.off())

gVlnSig <- VlnPlot(seuratMonoQC,
        features = "geneSetScore",
        split.by = "studyGroup",
        group.by = "seurat_clusters") +
  scale_fill_manual(values = c("darkcyan",
                               "red")) +
  labs(x = "cluster",
       y = "Bulk DE signature score")

png(file.path(plotDir,
              "Violin_MonocytesIntegrated_BulkDEScore.png"),
    height = 400,
    width = 800)

print(gVlnSig)

invisible(dev.off())

gBulkSelectedGenes <- FeaturePlot(seuratMonoQC,
            features = c("CD69",
                         "CSF1",
                         "C2",
                         "GSPT2",
                         "WNT5B",
                         "ZNF550"),
            reduction = "umap", 
            split.by = "studyGroup",
            min.cutoff = 'q1',
            max.cutoff = 'q99',
        raster = FALSE)&
  labs(x = "UMAP 1", y ="UMAP 2", color = "") &
  scale_color_viridis() &
  theme(aspect.ratio = 1)

png(file.path(plotDir,
              "UMAP_Monocytes_BulkSigGenes.png"),
    height = 800,
    width = 1200)

print(gBulkSelectedGenes)

invisible(dev.off())

gVlnSig <- VlnPlot(seuratMonoQC,
        features = c("CD69",
                         "CSF1",
                         "C2",
                         "GSPT2",
                         "WNT5B",
                         "ZNF550"),
        split.by = "studyGroup",
        group.by = "seurat_clusters") +
  scale_fill_manual(values = c("darkcyan",
                               "red")) +
  labs(x = "cluster")

png(file.path(plotDir,
              "Violin_MonocytesIntegrated_BulkSelectedGenes.png"),
    height = 800,
    width = 1200)

print(gVlnSig)

invisible(dev.off())



plotGeneViolin <- function(seuratIn,
                           selectedGene){
  
  gVlnSig <- VlnPlot(seuratIn,
        features = c(selectedGene),
        split.by = "studyGroup",
        group.by = "seurat_clusters") +
  scale_fill_manual(values = c("darkcyan",
                               "red")) +
  labs(x = "Cluster")
  
  filename <- paste0("Violin_Monocytes",selectedGene,".png")

png(file.path(plotDir,
              filename),
    height = 400,
    width = 800)

print(gVlnSig)

invisible(dev.off())

return(gVlnSig)
  
}


plotGeneViolin(seuratMonoQC,
               "IFITM1")

plotGeneViolin(seuratMonoQC,
               "S100A8")

plotGeneViolin(seuratMonoQC,
               "SLAMF7")
```

```{r mas_vs_hc_within_clusters}

Idents(seuratMonoQC) <- "seurat_clusters"
DefaultAssay(seuratMonoQC) <- "RNA"
MASVsHCByCellType <- NULL

cellTypes <- sort(unique(seuratMonoQC$seurat_clusters))
for(seurat_cluster in cellTypes){
  markersByClust <- FindMarkers(seuratMonoQC, 
                         ident.1 = "MAS", 
                         ident.2 = "HC", 
                         group.by = "studyGroup", 
                         subset.ident = seurat_cluster)
  markersByClust$gene <- rownames(markersByClust)
  markersByClust$cellType <- seurat_cluster
  
  MASVsHCByCellType[[seurat_cluster]] <- markersByClust
}

openxlsx::write.xlsx(MASVsHCByCellType, file = file.path(tableDir, "SeuratV5MASVsHCWithinClustersRes0_3.xlsx"))

MASVsHCByCellTypeLong <- bind_rows(MASVsHCByCellType)

```


```{r investigate_de_genes}
nGenes <- 2

hmapGenes <- MASVsHCByCellTypeLong %>%
  dplyr::group_by(cellType) %>%
  dplyr::filter(abs(avg_log2FC) > 0.5) %>%
  dplyr::slice_min(p_val_adj, n=nGenes) %>%
  dplyr::slice_max(avg_log2FC, n = nGenes) #%>%
  #dplyr::arrange(p_val_adj) %>%
  #dplyr::mutate(fc_order = order(p_val_adj))

selGenes <- unique(hmapGenes$gene)

DefaultAssay(seuratMonoQC) <-"RNA"

studyGroupColors <- c("darkcyan",
                               "red")

gVln <- VlnPlot(seuratMonoQC,
        feature = selGenes,
        pt.size = 0,
        split.by = "studyGroup",
        group.by = "seurat_clusters",
        ncol = 3) &
  scale_fill_manual(values = studyGroupColors)

png(file.path(plotDir,
              "Violin_V5_MASvsHC_DE.png"),
    height = 600,
    width = 1000)

print(gVln)

invisible(dev.off())

gVln <- VlnPlot(seuratMonoQC,
        feature = selGenes,
        pt.size = 0,
        split.by = "donorID",
        group.by = "seurat_clusters",
        ncol = 3) 

png(file.path(plotDir,
              "Violin_V5_MASvsHC_ByDonor_DE.png"),
    height = 600,
    width = 1000)

print(gVln)

invisible(dev.off())
```


```{r pseudo_bulk}

pseudoMono <- AggregateExpression(seuratMonoQC, assays = "RNA", return.seurat = T, group.by = c("donorID", "studyGroup", "seurat_clusters"))

Idents(pseudoMono) <- "seurat_clusters"
DefaultAssay(pseudoMono) <- "RNA"
MASVsHCByCellTypePseudo <- NULL

cellTypes <- sort(unique(pseudoMono$seurat_clusters))
for(seurat_cluster in cellTypes){
  markersByClust <- FindMarkers(pseudoMono, 
                         ident.1 = "MAS", 
                         ident.2 = "HC", 
                         group.by = "studyGroup", 
                         subset.ident = seurat_cluster)
  markersByClust$gene <- rownames(markersByClust)
  markersByClust$cellType <- seurat_cluster
  
  MASVsHCByCellTypePseudo[[seurat_cluster]] <- markersByClust
}

openxlsx::write.xlsx(MASVsHCByCellTypePseudo, file = file.path(tableDir, "SeuratV5MASVsHCPseudobulkClusterResultsAll.xlsx"))

MASVsHCByCellTypePseudoLong <- bind_rows(MASVsHCByCellTypePseudo)

nGenes <- 50

top50List <- MASVsHCByCellTypePseudoLong %>%
  dplyr::group_by(cellType) %>%
  dplyr::filter(abs(avg_log2FC) > 0.5) %>%
  dplyr::filter(gene != "MTRNR2L8") %>%
  dplyr::slice_min(p_val_adj, n=nGenes) %>%
  dplyr::slice_max(avg_log2FC, n = nGenes) %>%
  dplyr::rename(cluster = cellType) %>%
  group_split(cluster)

openxlsx::write.xlsx(top50List, file = file.path(tableDir, "SeuratV5MASVsHCPseudobulkClusterResultsTop50.xlsx"))
```


```{r investigate_de_genes_pseudobulk}
nGenes <- 5

hmapGenes <- MASVsHCByCellTypePseudoLong %>%
  dplyr::group_by(cellType) %>%
  dplyr::filter(abs(avg_log2FC) > 0.5) %>%
  dplyr::filter(gene != "MTRNR2L8") %>%
  dplyr::slice_min(p_val_adj, n=nGenes) %>%
  dplyr::slice_max(avg_log2FC, n = nGenes) #%>%
  #dplyr::arrange(p_val_adj) %>%
  #dplyr::mutate(fc_order = order(p_val_adj))

selGenes <- unique(hmapGenes$gene)

# selGenes <- c("IFI27",
#               "IFITM1",
#               "S100A8",
#               "C1QB")

DefaultAssay(seuratMonoQC) <-"RNA"

studyGroupColors <- c("darkcyan",
                               "red")

# gVln <- VlnPlot(seuratMonoQC,
#         feature = selGenes,
#         pt.size = 0,
#         split.by = "studyGroup",
#         group.by = "seurat_clusters",
#         ncol = 3) &
#   scale_fill_manual(values = studyGroupColors)
# 
# # png(file.path(plotDir,
# #               "Violin_V5_MASvsHC_PseudobulkDE.png"),
# #     height = 600,
# #     width = 1000)
# # 
# # print(gVln)
# # 
# # invisible(dev.off())
# 
# gVln <- VlnPlot(seuratMonoQC,
#         feature = selGenes,
#         pt.size = 0,
#         split.by = "donorID",
#         group.by = "seurat_clusters",
#         ncol = 2) 
# 
# png(file.path(plotDir,
#               "Violin_V5_MASvsHC_ByDonorPseudobulk_DE.png"),
#     height = 600,
#     width = 1000)
# 
# print(gVln)
# 
# invisible(dev.off())
# 
# library(DoMultiBarHeatmap)
# 
# DoMultiBarHeatmap(pseudoMono,
#           features = selGenes,
#           group.by='seurat_clusters', additional.group.by = 'donorID')

library(Scillus)

donorColors <- rcartocolor::carto_pal(12, "Vivid")
#donorColors <- c(rep("red", 8), rep("darkcyan", 4))

pdf(file.path(plotDir, "HeatmapTopPseudobulkDE.pdf"),
    height = 6,
    width = 10,
    onefile = F)

plot_heatmap(dataset = pseudoMono, 
              markers = selGenes,
              sort_var = c("seurat_clusters","studyGroup","donorID"),
              anno_var = c("seurat_clusters","studyGroup","donorID"),
              anno_colors = list(clusterColors,
                                 studyGroupColors,
                                 donorColors))

invisible(dev.off())
```


# Lymphocyte analysis

```{r subset_lymphocytes}

DimPlot(seuratLymphQC, reduction = "umap", group.by = c("predicted.celltype.l1")) 

seuratLymphQC$UMAP1 <- seuratLymphQC[["umap"]]@cell.embeddings[,1]
seuratLymphQC$UMAP2 <- seuratLymphQC[["umap"]]@cell.embeddings[,2]

seuratLymphQC <- subset(seuratLymphQC,
                          subset = (UMAP2 < 10))

seuratLymphQC <- FindNeighbors(seuratLymphQC, reduction = "integrated.rpca", dims = 1:30)
seuratLymphQC <- FindClusters(seuratLymphQC, resolution = 0.3)

seuratLymphQC <- RunUMAP(seuratLymphQC, dims = 1:30, reduction = "integrated.rpca")

saveRDS(seuratLymphQC, file.path(dataDir, "SeuratV5LymphQC.RDS"))

seuratLymphQC <- readRDS(file.path(dataDir, "SeuratV5LymphQC.RDS"))

```

```{r plot_lymphocyte_umaps}

png(file.path(plotDir, "UMAPLymphocytesClustersRes0_3.png"),
    height = 400,
    width = 600)

DimPlot(seuratLymphQC, reduction = "umap", 
        group.by = c("seurat_clusters"),
        label = TRUE)+
  labs(x = "UMAP 1",
       y = "UMAP 2")+
  theme(aspect.ratio = 1)

invisible(dev.off())

png(file.path(plotDir, "UMAPLymphocytesPredictedCelltypeL1.png"),
    height = 400,
    width = 600)

DimPlot(seuratLymphQC, reduction = "umap", 
        group.by = c("predicted.celltype.l1"),
        label = TRUE)+
  labs(x = "UMAP 1",
       y = "UMAP 2")+
  theme(aspect.ratio = 1)

invisible(dev.off())

png(file.path(plotDir, "UMAPLymphocytesPredictedCelltypeL2.png"),
    height = 400,
    width = 600)

DimPlot(seuratLymphQC, reduction = "umap", 
        group.by = c("predicted.celltype.l2"),
        label = TRUE)+
  labs(x = "UMAP 1",
       y = "UMAP 2")+
  theme(aspect.ratio = 1)

invisible(dev.off())

```

```{r export_mono_for_rnavelo}

# save metadata table:
seuratMonoQC$barcode <- colnames(seuratMonoQC)
seuratMonoQC$UMAP_1 <- seuratMonoQC@reductions$umap@cell.embeddings[,1]
seuratMonoQC$UMAP_2 <- seuratMonoQC@reductions$umap@cell.embeddings[,2]
write.csv(seuratMonoQC@meta.data, 
          file=file.path(dataDir,
                         "MonocyteMetadataForRNAvelo.csv"), quote=F, row.names=F)

# write expression counts matrix
library(Matrix)
counts_matrix <- GetAssayData(seuratMonoQC, assay='RNA', slot='counts')
writeMM(counts_matrix, file=file.path(dataDir,'CountsForRNAvelo.mtx'))

# write dimesnionality reduction matrix, in this example case pca matrix
write.csv(seuratMonoQC@reductions$pca@cell.embeddings, 
          file=file.path(dataDir,"PCAForRNAvelo.csv"), quote=F, row.names=F)

# write gene names
write.table(
  data.frame('gene'=rownames(counts_matrix)),file=file.path(dataDir, "GeneNamesForRNAvelo.csv"),
  quote=F,row.names=F,col.names=F
)

#Check barcode ends for comparison with samples

seuratMonoQC$barcodeEnds <- str_extract(seuratMonoQC$barcode, "[0-9]+_[0-9]+_[0-9]+")
table(seuratMonoQC$sample, seuratMonoQC$barcodeEnds)

```


```{r export_for_loupe}

library(loupeR)

#Check assay
DefaultAssay(seuratLymphQC)
DefaultAssay(seuratMonoQC)

create_loupe_from_seurat(
    seuratLymphQC,
    output_dir = dataDir,
    output_name = "LymphocyteLoupeFile",
    force = TRUE)

create_loupe_from_seurat(
    seuratMonoQC,
    output_dir = dataDir,
    output_name = "MonocyteLoupeFile",
    force = TRUE)

```

```{r downsample_export_for_loupe}

#Check assay
DefaultAssay(seuratLymphQC)
DefaultAssay(seuratMonoQC)

#Set identity class for downsampling
Idents(seuratMonoQC) <- "studyGroup"

dsNMonocytes <- 20000
seuratMonoDS <- subset(seuratMonoQC, downsample = dsNMonocytes)

create_loupe_from_seurat(
    seuratMonoDS,
    output_dir = dataDir,
    output_name = "DownsampledMonocyteLoupeFile",
    force = TRUE)

#Set identity class for downsampling
Idents(seuratLymphQC) <- "studyGroup"

dsNLymphocytes <- 10000
seuratLymphDS <- subset(seuratLymphQC, downsample = dsNLymphocytes)

create_loupe_from_seurat(
    seuratLymphDS,
    output_dir = dataDir,
    output_name = "DownsampledLymphocyteLoupeFile",
    force = TRUE)

```

```{r combine_for_cellphoneDB}

#Merge lymphocytes and monocytes

seuratAllInt <- merge(seuratLymphQC,
                   seuratMonoQC)

seuratAllInt@assays$RNA$scale.data.1 <- NULL
seuratAllInt@assays$RNA$scale.data.2 <- NULL

#Convert to V3 object
seuratAllInt[["RNA"]] <- as(object = seuratAllInt[["RNA"]], Class = "Assay")

seuratAllInt$cellType <- case_when(str_detect(seuratAllInt$sampleName, "[m,M]ono") ~ "monocytes",
                                   str_detect(seuratAllInt$sampleName, "[l,L]ymph") ~ "lymphocytes")


metadataCPDB <- seuratAllInt@meta.data %>%
  dplyr::mutate(cell_type = case_when(cellType == "lymphocytes" ~ predicted.celltype.l2,
                                      cellType == "monocytes" ~ seurat_clusters),
                barcode = colnames(seuratAllInt)) %>%
  dplyr::select(barcode, cell_type) %>%
  dplyr::rename(barcode_sample = barcode)

seuratAllInt$cell_type = metadataCPDB$cell_type

library(SeuratDisk)
SaveH5Seurat(
  seuratAllInt, 
  filename = 
    file.path(dataDir, "seuratLymphocytesAndMonocytesV5Unscaled.h5Seurat"),
  overwrite = TRUE)
Convert(
  source = file.path(dataDir,"seuratLymphocytesAndMonocytesV5Unscaled.h5Seurat"),
  dest = "h5ad", 
  assay = "RNA",
  overwrite = TRUE)


write.table(metadataCPDB,
          file = file.path(dataDir, "seuratLymphocytesAndMonocytesV520240301.tsv"),
          sep = "\t")

```

```{r combine_and_save_all}

seuratMonoQC$sort <- "Monocyte"
seuratLymphQC$sort <- "Lymphocyte"

seuratMonoQC$sortCluster <- paste0("Mono_", seuratMonoQC$seurat_clusters)
seuratLymphQC$sortCluster <- paste0("Lymph_", seuratLymphQC$seurat_clusters)

seuratAll <- merge(seuratLymphQC,
                   seuratMonoQC)

saveRDS(seuratAll, file.path(dataDir, "SeuratV5All.RDS"))
```
